File: inferlet/src/interface/forward.rs
use crate::brle::Brle;
use crate::forward;
use crate::{Queue, Resource};
use std::rc::Rc;
use wstd::io::AsyncPollable;

#[derive(Debug, Clone)]
pub struct ForwardPass {
    pub(crate) inner: Rc<forward::ForwardPass>,
}

#[derive(Debug, Clone)]
pub struct ForwardPassResult {
    pub distributions: Option<Vec<Distribution>>,
    pub tokens: Option<Vec<u32>>,
}

/// Represents a probability distribution over a set of tokens.
#[derive(Clone, Debug)]
pub struct Distribution {
    /// A vector of token IDs.
    pub ids: Vec<u32>,
    /// A vector of probabilities corresponding to the token IDs.
    pub probs: Vec<f32>,
}

// "Smart" kv page
#[derive(Debug, Clone)]
pub struct KvPage {
    queue: Queue,
    rc: Rc<()>,
    ptr: u32,
}

impl KvPage {
    pub fn new(queue: &Queue, ptr: u32) -> Self {
        KvPage {
            queue: queue.clone(),
            rc: Rc::new(()),
            ptr,
        }
    }

    pub fn ptr(&self) -> u32 {
        self.ptr
    }
}

impl Drop for KvPage {
    fn drop(&mut self) {
        if Rc::strong_count(&self.rc) == 1 {
            self.queue.deallocate_kv_page_ptr(self.ptr);
        }
    }
}

pub fn causal_mask(num_total_tokens: u32, num_input_tokens: u32) -> Vec<Brle> {
    let mut mask = Vec::new();
    let offset = num_total_tokens - num_input_tokens;
    for i in 0..num_input_tokens {
        mask.push(Brle::new((offset + i + 1) as usize));
    }
    mask
}

/// Defines the interface for executing forward passes in a neural network model,
/// including support for PEFT adapters and KV cache manipulation.
pub trait Forward {
    fn new_kv_page(&self) -> KvPage;
    fn new_kv_pages(&self, count: usize) -> Vec<KvPage>;

    fn export_kv_pages(&self, ptrs: &[KvPage], name: &str);
    fn import_kv_pages(&self, name: &str) -> Vec<KvPage>;
    fn allocate_kv_page_ptr(&self) -> u32;
    fn allocate_kv_page_ptrs(&self, count: usize) -> Vec<u32>;
    fn deallocate_kv_page_ptr(&self, ptr: u32);
    fn deallocate_kv_page_ptrs(&self, ptrs: &[u32]);
    fn export_kv_page_ptrs(&self, ptrs: &[u32], name: &str);
    fn import_kv_page_ptrs(&self, name: &str) -> Vec<u32>;

    fn get_all_exported_kv_pages(&self) -> Vec<(String, u32)>;
    fn release_exported_kv_pages(&self, name: &str);

    fn allocate_embed_ptr(&self) -> u32;
    fn allocate_embed_ptrs(&self, count: usize) -> Vec<u32>;
    fn deallocate_embed_ptr(&self, ptr: u32);
    fn deallocate_embed_ptrs(&self, ptrs: &[u32]);
    fn export_embed_ptrs(&self, ptrs: &[u32], name: &str);
    fn import_embed_ptrs(&self, name: &str) -> Vec<u32>;
    fn get_all_exported_embeds(&self) -> Vec<(String, u32)>;
    fn release_exported_embeds(&self, name: &str);

    fn create_forward_pass(&self) -> ForwardPass;
}

impl Forward for Queue {
    fn new_kv_page(&self) -> KvPage {
        let ptr = self.allocate_kv_page_ptr();
        KvPage::new(self, ptr)
    }

    fn new_kv_pages(&self, count: usize) -> Vec<KvPage> {
        self.allocate_kv_page_ptrs(count)
            .into_iter()
            .map(|ptr| KvPage::new(self, ptr))
            .collect()
    }

    fn export_kv_pages(&self, kv_pages: &[KvPage], name: &str) {
        let ptrs = kv_pages.iter().map(|kv| kv.ptr()).collect::<Vec<_>>();
        self.export_resource(Resource::KvPage, &ptrs, name)
    }

    fn import_kv_pages(&self, name: &str) -> Vec<KvPage> {
        let ptrs = self.import_resource(Resource::KvPage, name);
        ptrs.into_iter().map(|ptr| KvPage::new(self, ptr)).collect()
    }

    fn allocate_kv_page_ptr(&self) -> u32 {
        self.allocate_resources(Resource::KvPage, 1)
            .into_iter()
            .next()
            .unwrap()
    }

    fn allocate_kv_page_ptrs(&self, count: usize) -> Vec<u32> {
        self.allocate_resources(Resource::KvPage, count as u32)
    }

    fn deallocate_kv_page_ptr(&self, ptr: u32) {
        self.deallocate_resources(Resource::KvPage, &[ptr])
    }

    fn deallocate_kv_page_ptrs(&self, ptrs: &[u32]) {
        self.deallocate_resources(Resource::KvPage, ptrs)
    }

    fn export_kv_page_ptrs(&self, ptrs: &[u32], name: &str) {
        self.export_resource(Resource::KvPage, ptrs, name)
    }

    fn import_kv_page_ptrs(&self, name: &str) -> Vec<u32> {
        self.import_resource(Resource::KvPage, name)
    }

    fn get_all_exported_kv_pages(&self) -> Vec<(String, u32)> {
        self.get_all_exported_resources(Resource::KvPage)
    }

    fn release_exported_kv_pages(&self, name: &str) {
        self.release_exported_resources(Resource::KvPage, name)
    }

    fn allocate_embed_ptr(&self) -> u32 {
        self.allocate_resources(Resource::Embed, 1)
            .into_iter()
            .next()
            .unwrap()
    }

    fn allocate_embed_ptrs(&self, count: usize) -> Vec<u32> {
        self.allocate_resources(Resource::Embed, count as u32)
    }

    fn deallocate_embed_ptr(&self, ptr: u32) {
        self.deallocate_resources(Resource::Embed, &[ptr])
    }

    fn deallocate_embed_ptrs(&self, ptrs: &[u32]) {
        self.deallocate_resources(Resource::Embed, ptrs)
    }

    fn export_embed_ptrs(&self, ptrs: &[u32], name: &str) {
        self.export_resource(Resource::Embed, ptrs, name)
    }

    fn import_embed_ptrs(&self, name: &str) -> Vec<u32> {
        self.import_resource(Resource::Embed, name)
    }

    fn get_all_exported_embeds(&self) -> Vec<(String, u32)> {
        self.get_all_exported_resources(Resource::Embed)
    }

    fn release_exported_embeds(&self, name: &str) {
        self.release_exported_resources(Resource::Embed, name)
    }

    fn create_forward_pass(&self) -> ForwardPass {
        ForwardPass {
            inner: Rc::new(forward::create_forward_pass(&self.inner)),
        }
    }
}

impl ForwardPass {
    pub async fn execute(&self) -> ForwardPassResult {
        if let Some(future) = self.inner.execute() {
            let pollable = future.pollable();
            AsyncPollable::new(pollable).wait_for().await;

            let mut dists = Vec::new();
            if let Some(distributions) = future.get_distributions() {
                for (ids, probs) in distributions {
                    dists.push(Distribution { ids, probs });
                }
            }
            let distributions = if dists.is_empty() { None } else { Some(dists) };

            ForwardPassResult {
                distributions,
                tokens: future.get_tokens(),
            }
        } else {
            ForwardPassResult {
                distributions: None,
                tokens: None,
            }
        }
    }

    pub fn input_embed_ptrs(&self, embed_ptrs: &[u32], positions: &[u32]) {
        forward::input_embeddings(&self.inner, embed_ptrs, positions);
    }

    pub fn input_tokens(&self, input_tokens: &[u32], positions: &[u32]) {
        forward::input_tokens(&self.inner, input_tokens, positions);
    }

    pub fn output_embed_ptrs(&self, embed_ptrs: &[u32], indices: &[u32]) {
        forward::output_embeddings(&self.inner, embed_ptrs, indices);
    }

    pub fn output_distributions(&self, indices: &[u32], temperature: f32, top_k: Option<u32>) {
        forward::output_distributions(&self.inner, indices, temperature, top_k);
    }

    pub fn output_tokens(&self, indices: &[u32], temperature: f32) {
        forward::output_tokens(&self.inner, indices, temperature);
    }

    pub fn output_tokens_top_p(&self, indices: &[u32], temperature: f32, top_p: f32) {
        forward::output_tokens_top_p(&self.inner, indices, temperature, top_p);
    }

    pub fn output_tokens_top_k(&self, indices: &[u32], temperature: f32, top_k: u32) {
        forward::output_tokens_top_k(&self.inner, indices, temperature, top_k);
    }

    pub fn output_tokens_min_p(&self, indices: &[u32], temperature: f32, min_p: f32) {
        forward::output_tokens_min_p(&self.inner, indices, temperature, min_p);
    }

    pub fn output_tokens_top_k_top_p(
        &self,
        indices: &[u32],
        temperature: f32,
        top_k: u32,
        top_p: f32,
    ) {
        forward::output_tokens_top_k_top_p(&self.inner, indices, temperature, top_k, top_p);
    }

    pub fn attention_mask(&self, mask: &[Vec<u32>]) {
        forward::attention_mask(&self.inner, mask);
    }

    pub fn kv_cache(&self, kv_pages: &[KvPage], last_kv_page_len: usize) {
        let ptrs = kv_pages.iter().map(|kv| kv.ptr()).collect::<Vec<_>>();
        forward::kv_cache(&self.inner, &ptrs, last_kv_page_len as u32);
    }

    pub fn kv_cache_ptrs(&self, kv_page_ptrs: &[u32], last_kv_page_len: usize) {
        forward::kv_cache(&self.inner, kv_page_ptrs, last_kv_page_len as u32);
    }
}

File: inferlet-js/src/forward.ts
// ForwardPass, KvPage, and Distribution types for neural network forward passes.
// Mirrors the Rust forward.rs from inferlet/src/forward.rs

import type { Queue } from './model.js';
import { Brle } from './brle.js';

// WIT bindings - these will be available at runtime through componentize-js
import type { ForwardPass as WitForwardPass } from 'inferlet:core/forward';
import * as forward from 'inferlet:core/forward';

/**
 * Represents a probability distribution over a set of tokens.
 */
export interface Distribution {
  /** Token IDs */
  ids: number[];
  /** Probabilities corresponding to the token IDs */
  probs: number[];
}

/**
 * Result of executing a forward pass.
 */
export interface ForwardPassResult {
  /** Optional distributions for requested output indices */
  distributions?: Distribution[];
  /** Optional sampled tokens for requested output indices */
  tokens?: number[];
}

/**
 * Represents a KV cache page with automatic resource management.
 * When the last reference is dropped, the page is deallocated.
 */
export class KvPage {
  private _queue: Queue;
  private _ptr: number;
  private _released: boolean = false;

  constructor(queue: Queue, ptr: number) {
    this._queue = queue;
    this._ptr = ptr;
  }

  /**
   * Returns the raw pointer to this KV page.
   */
  get ptr(): number {
    return this._ptr;
  }

  /**
   * Explicitly release this KV page.
   * Called automatically when the page is no longer needed.
   */
  release(): void {
    if (!this._released) {
      this._queue.deallocateKvPagePtr(this._ptr);
      this._released = true;
    }
  }
}

/**
 * ForwardPass represents a batch of forward pass operations.
 * Configure inputs, outputs, attention masks, and KV cache before executing.
 */
export class ForwardPass {
  private inner: WitForwardPass;

  constructor(inner: WitForwardPass) {
    this.inner = inner;
  }

  /**
   * Execute the forward pass asynchronously.
   */
  async execute(): Promise<ForwardPassResult> {
    const future = this.inner.execute();
    if (!future) {
      return { distributions: undefined, tokens: undefined };
    }

    // Wait for the async operation to complete
    const pollable = future.pollable();
    // In WASM context, wstd handles the polling
    await new Promise<void>((resolve) => {
      const checkPoll = () => {
        if (pollable.ready()) {
          resolve();
        } else {
          setTimeout(checkPoll, 0);
        }
      };
      checkPoll();
    });

    // Get results
    const distributions: Distribution[] = [];
    const rawDistributions = future.getDistributions();
    if (rawDistributions) {
      for (const [ids, probs] of rawDistributions) {
        distributions.push({ ids: [...ids], probs: [...probs] });
      }
    }

    const tokens = future.getTokens();

    return {
      distributions: distributions.length > 0 ? distributions : undefined,
      tokens: tokens ? [...tokens] : undefined,
    };
  }

  /**
   * Set embedding pointers as input.
   */
  inputEmbedPtrs(embedPtrs: number[], positions: number[]): void {
    forward.inputEmbeddings(this.inner, embedPtrs, positions);
  }

  /**
   * Set token IDs as input.
   */
  inputTokens(inputTokens: number[], positions: number[]): void {
    forward.inputTokens(this.inner, inputTokens, positions);
  }

  /**
   * Set embedding pointers for output capture.
   */
  outputEmbedPtrs(embedPtrs: number[], indices: number[]): void {
    forward.outputEmbeddings(this.inner, embedPtrs, indices);
  }

  /**
   * Request probability distributions at specified indices.
   */
  outputDistributions(indices: number[], temperature: number, topK?: number): void {
    forward.outputDistributions(this.inner, indices, temperature, topK);
  }

  /**
   * Request sampled tokens at specified indices (multinomial sampling).
   */
  outputTokens(indices: number[], temperature: number): void {
    forward.outputTokens(this.inner, indices, temperature);
  }

  /**
   * Request sampled tokens using top-p (nucleus) sampling.
   */
  outputTokensTopP(indices: number[], temperature: number, topP: number): void {
    forward.outputTokensTopP(this.inner, indices, temperature, topP);
  }

  /**
   * Request sampled tokens using top-k sampling.
   */
  outputTokensTopK(indices: number[], temperature: number, topK: number): void {
    forward.outputTokensTopK(this.inner, indices, temperature, topK);
  }

  /**
   * Request sampled tokens using min-p sampling.
   */
  outputTokensMinP(indices: number[], temperature: number, minP: number): void {
    forward.outputTokensMinP(this.inner, indices, temperature, minP);
  }

  /**
   * Request sampled tokens using combined top-k and top-p sampling.
   */
  outputTokensTopKTopP(
    indices: number[],
    temperature: number,
    topK: number,
    topP: number
  ): void {
    forward.outputTokensTopKTopP(this.inner, indices, temperature, topK, topP);
  }

  /**
   * Set the attention mask for the forward pass.
   * Each element is a Brle or raw buffer representing which positions are visible.
   */
  attentionMask(mask: Brle[] | number[][]): void {
    // Convert Brle array to raw buffers if needed
    const rawMask = mask.map((m) => (m instanceof Brle ? m.buffer : m));
    forward.attentionMask(this.inner, rawMask);
  }

  /**
   * Set the KV cache for the forward pass.
   */
  kvCache(kvPages: KvPage[], lastKvPageLen: number): void {
    const ptrs = kvPages.map((kv) => kv.ptr);
    forward.kvCache(this.inner, ptrs, lastKvPageLen);
  }

  /**
   * Set the KV cache using raw pointers.
   */
  kvCachePtrs(kvPagePtrs: number[], lastKvPageLen: number): void {
    forward.kvCache(this.inner, kvPagePtrs, lastKvPageLen);
  }
}

/**
 * Resource type enumeration for allocation/deallocation.
 */
export enum Resource {
  KvPage = 0,
  Embed = 1,
  Adapter = 2,
}

/**
 * Interface for forward pass operations.
 * Implemented by Queue.
 */
export interface Forward {
  // KvPage management with smart pointers
  newKvPage(): KvPage;
  newKvPages(count: number): KvPage[];

  // KvPage raw pointer management
  allocateKvPagePtr(): number;
  allocateKvPagePtrs(count: number): number[];
  deallocateKvPagePtr(ptr: number): void;
  deallocateKvPagePtrs(ptrs: number[]): void;

  // KvPage export/import
  exportKvPages(kvPages: KvPage[], name: string): void;
  importKvPages(name: string): KvPage[];
  exportKvPagePtrs(ptrs: number[], name: string): void;
  importKvPagePtrs(name: string): number[];
  getAllExportedKvPages(): [string, number][];
  releaseExportedKvPages(name: string): void;

  // Embedding pointer management
  allocateEmbedPtr(): number;
  allocateEmbedPtrs(count: number): number[];
  deallocateEmbedPtr(ptr: number): void;
  deallocateEmbedPtrs(ptrs: number[]): void;
  exportEmbedPtrs(ptrs: number[], name: string): void;
  importEmbedPtrs(name: string): number[];
  getAllExportedEmbeds(): [string, number][];
  releaseExportedEmbeds(name: string): void;

  // ForwardPass creation
  createForwardPass(): ForwardPass;
}

// Re-export causalMask from brle
export { causalMask } from './brle.js';
